projects/style_gen/modules.py

```diff
@@ -57,3 +71,319 @@ def get_temp_history(self, observation: Message) -> Optional[str]:
+            tensor = _normalize(tensor, self.norm_embeddings)
+
+        return tensor, new_incr_state
+
+
+class ClassificationMixin(Agent):
+    """
+    Mixin for adding classification metrics to non-classifier models.
+    """
+
+    def __init__(self, opt: Opt, shared=None):
+        super().__init__(opt, shared)
+        if shared is None:
+            self.reset_metrics()
+
+    def _update_confusion_matrix(
+        self, predictions: List[Optional[str]], labels: List[Optional[List[str]]]
+    ):
+        """
+        Update the confusion matrix given the batch and predictions.
+
+        :param predictions: List of strings of length batchsize. Each string is a label
+            predicted by the classifier. A string will be None if the corresponding
+            observation is empty.
+        :param labels: List of label fields from the observations. Fields may be Nones
+            if the observations are empty.
+        """
+        f1_dict = {}
+        explode_labels = []
+        for x in labels:
+            if x is not None and len(x) > 0:
+                assert len(x) == 1, 'Multiple labels are not currently supported!'
+                explode_labels.append(x[0])
+            else:
+                explode_labels.append(None)
+
+        # Check that predictions and labels have Nones in the same places, and then
+        # filter the Nones out because we can't compute metrics with them
+        assert len(predictions) == len(labels)
+        assert all(
+            [
+                (pred is None and label is None)
+                or (pred is not None and label is not None)
+                for pred, label in zip(predictions, explode_labels)
+            ]
+        )
+        filtered_predictions = [pred for pred in predictions if pred is not None]
+        filtered_labels = [label for label in explode_labels if label is not None]
+
+        class_list = set(filtered_predictions + filtered_labels)
+        for class_name in class_list:
+            prec_str = f'class_{class_name}_prec'
+            recall_str = f'class_{class_name}_recall'
+            f1_str = f'class_{class_name}_f1'
+            precision, recall, f1 = ConfusionMatrixMetric.compute_metrics(
+                filtered_predictions, filtered_labels, class_name
+            )
+            f1_dict[class_name] = f1
+            self.record_local_metric(prec_str, precision)
+            self.record_local_metric(recall_str, recall)
+            self.record_local_metric(f1_str, f1)
+        self.record_local_metric('weighted_f1', WeightedF1Metric.compute_many(f1_dict))
+
+    def _get_preds(self, batch_reply):
+        preds = [reply.get('text') for reply in batch_reply]
+        if all(x is None for x in preds):
+            return None
+
+        return preds
+
+    def _get_labels(self, observations, labels_field: str):
+        labels = [obs.get(labels_field) for obs in observations]
+        return labels
+
+    def batch_act(self, observations):
```

Comment:
is there a difference between this and `TorchAgent.batch_act`? just curious

Reply:
Hmm, there isn't anymore, due to a fix that Candace made a few days ago. I've deleted this `batch_act()` because it's no longer useful